File: others/2000-2100/2035. 将数组分成两个数组并最小化数组和的差.py
from collections import *
from functools import *
from math import *
from itertools import *
from heapq import *
from bisect import *
from typing import *
from queue import *
from string import *

DIRS = [(1, 1), (1, -1), (-1, -1), (-1, 1)]
MOD = int(1e9 + 7)


class Solution:
    def minimumDifference(self, nums: List[int]) -> int:
        n = len(nums) // 2
        total = sum(nums)

        def count(nums: List[int]) -> DefaultDict:
            ret = defaultdict(set)
            for i in range(1 << len(nums)):
                cnt = 0
                s = 0
                for j, x in enumerate(nums):
                    if i >> j & 1:
                        s += x
                        cnt += 1
                ret[cnt].add(s)
            return ret

        # 快一点
        def count(nums: List[int]) -> DefaultDict:
            ret = defaultdict(set)
            n = len(nums)
            for k in range(n + 1):
                for comb in combinations(nums, k):
                    ret[k].add(sum(comb))
            return ret

        left = count(nums[:n])
        right = count(nums[n:])
        for k in right:
            right[k] = sorted(right[k])
        ans = inf
        for k, left_sums in left.items():
            right_sums = right[n - k]
            for s1 in left_sums:
                target = total // 2 - s1
                idx = bisect_left(right_sums, target)
                # 检查 idx 和 idx - 1（防止漏最接近的）
                for j in [idx - 1, idx]:
                    if 0 <= j < len(right_sums):
                        s2 = right_sums[j]
                        s = s1 + s2
                        ans = min(ans, abs(total - 2 * s))
        return ans

File: leetcode/2000-2100/2035. 将数组分成两个数组并最小化数组和的差.cpp
#include <bits/stdc++.h>
using namespace std;
vector<vector<int>> DIRS = { {1, 1}, {1, -1}, {-1, -1}, {-1, 1} };
constexpr int MOD = int(1e9 + 7);


class Solution {
public:
    int minimumDifference(vector<int> &nums) {
        int total = reduce(nums.begin(), nums.end(), 0);
        int half = total / 2;

        auto count = [&](const vector<int> &nums) -> unordered_map<int, vector<int>> {
            unordered_map<int, vector<int>> ret;
            for (int i = 0; i < 1 << nums.size(); ++i) {
                int cnt = 0;
                int s = 0;
                for (int j = 0; j < nums.size(); ++j) {
                    if (i >> j & 1) {
                        cnt += 1;
                        s += nums[j];
                    }
                }
                ret[cnt].emplace_back(s);
            }
            return ret;
            };
        int n = nums.size();
        auto left = count({ nums.begin(), nums.begin() + n / 2 });
        auto right = count({ nums.begin() + n / 2, nums.end() });
        for (auto &[k, v] : right) {
            ranges::sort(v);
        }
        int m = n / 2;
        int ans = INT_MAX;
        for (auto &[k, v] : left) {
            for (int x : v) {
                auto it = ranges::lower_bound(right[m - k], half - x);
                if (it != right[m - k].end()) {
                    ans = min(ans, abs(2 * (x + *it) - total));
                    if (it != right[m - k].begin()) {
                        ans = min(ans, abs(2 * (x + *(prev(it))) - total));
                    }
                }
            }
        }
        return ans;
    }
};